x-pack/plugin/esql/src/main/antlr/EsqlBaseParser.g4

```diff
@@ -78,7 +78,7 @@ regexBooleanExpression
     ;
 
 matchBooleanExpression
-    : fieldExp=qualifiedName COLON queryString=constant
+    : fieldExp=qualifiedName COLON matchQuery=constant (CAST_OP dataType)?
```

Comment:
I've allowed to use casting on a constant. 

This is redundant however, as using directly a string does the right thing. The following are equivalent:

```
WHERE match(ip_field, "127.0.0.1"::IP)

WHERE match(ip_field, "127.0.0.1")
```

We can keep the casting for extra type safety, but otherwise it is unnecessary.

Reply:
> Is my understanding correct that match always works with literal strings no matter what the field data type is?

That is correct.

>  If this is true, adding explicit casting does seem unnecessary, to be honest, I'm confused by : + :: when I saw it the first time.

That was my first reaction as well - I think it can make sense for these use cases:
- Users are used to do casting for data types like IP
- Users want to validate the type is correct - match doesn't do that out of the box, but returns empty results for strings.

It's true that the match operator doesn't accept the to_XXX functions and we just provide casting for this use case, which can be weird as well.

I'm happy to remove the casting option if you feel this is confusing. I'd like to ping @ioanatia on this as she suggested using casting for opinions here.

Reply:
In case we want to support multiple field types, we may have something like the following:

`WHERE my_field::int : 32`

So maybe we want to provide a similar casting option for the query value so they both can use casting:

`WHERE my_field : "192.168.1.1"::ip`

besides that, and type safety, I can't think of a valid reason to provide casting on the query value.

@ioanatia , what are your thoughts here, as you suggested adding casting to the query value?